### PR TITLE
[server] Propagate leadership termId to all messages + follower-side stale filtering

### PR DESCRIPTION
**Pull request by sushantmane**
## Summary

Fixes the **Delayed Leadership Problem (DLP)** where a stale leader produces to the version topic after a new leader has taken over, causing data rollback and replica divergence.

- **Part 1 — termId propagation (always-on):** Stamps standalone leader-produced records and chunked-value manifests with the replica's own leadership term (`activeLeaderTerm` from the Helix state transition timestamp), replacing the hardcoded `DEFAULT_TERM_ID` (-1) on the produce paths:
  - `produceToLocalKafka` — used for both non-chunked records and the manifest of chunked records
  - `propagateHeartbeatFromUpstreamTopicToLocalVersionTopic`
  - `sendGlobalRtDivMessage`

  Intermediate chunk messages produced by `VeniceWriter.putLargeValue` continue to carry `DEFAULT_TERM_ID` and are intentionally not filtered (see "Chunked records" below).
- **Part 2 — Follower-side filtering** (gated by `server.leader.term.filtering.enabled`, default `false`): Followers reject data records whose `termId` is strictly less than the highest `termId` observed so far. Records with `DEFAULT_TERM_ID` (-1) are never filtered for backward compatibility during mixed rollout and to allow chunk messages to flow through. Control messages (DoL stamps, SOS, EOS, heartbeats) are unaffected. When a record is filtered, only the local VT position is advanced (so the drainer moves past) — upstream RT / remote-VT positions and leader GUID/host are intentionally left untouched so a stale leader's record cannot corrupt offset-rewind detection or split-brain tracking.
- **Metrics:** `ingestion.stale_leader_records_filtered.count` — OpenTelemetry `COUNTER` keyed by store-name + cluster-name + version-role. Operators alert and drill in via OTel dimensions.

### Chunked records

For values large enough to be chunked by `VeniceWriter.putLargeValue`:
- The chunk messages themselves carry `DEFAULT_TERM_ID` (sent via `DEFAULT_LEADER_METADATA_WRAPPER`). This is preexisting writer behavior and remains unchanged in this PR.
- The chunked-value **manifest** carries the leader's real `activeLeaderTerm`.

Since a follower can only assemble a value once it has both the chunks and the manifest, filtering a stale manifest is sufficient to prevent the stale value from ever being read — the chunks become orphaned bytes in storage but are never assembled or returned. Plumbing `termId` through chunk messages so that chunks and their manifest are filtered together (cleaner, no orphans) is a reasonable follow-up and tracked under "Follow-up" below.

### Files changed (12)

| File | Change |
|------|--------|
| `PartitionConsumptionState.java` | New `activeLeaderTerm` field (separate from `highestLeadershipTerm` to prevent cross-contamination during filtering) |
| `LeaderFollowerStoreIngestionTask.java` | Set `activeLeaderTerm` in S→L; clear it on L→S only after the leader-topic drain completes; replace `DEFAULT_TERM_ID` on all three leader-produce paths |
| `StoreIngestionTask.java` | `shouldFilterStaleLeaderRecord()` + integration in data-message branch; filtered records advance only the local VT position (skip upstream/leader-identity mutation); thread-safety doc; debug log on term update |
| `ConfigKeys.java` | `SERVER_LEADER_TERM_FILTERING_ENABLED` (static config, requires restart) |
| `VeniceServerConfig.java` | Parse + getter |
| `IngestionOtelMetricEntity.java` | New `STALE_LEADER_RECORDS_FILTERED_COUNT` OTel COUNTER entity |
| `IngestionOtelStats.java` / `NoOpIngestionOtelStats.java` | OTel sensor + `recordStaleLeaderRecordsFilteredCount` |
| `AggVersionedIngestionStats.java` | Delegation method that records to OTel only |
| `PartitionConsumptionStateTest.java` | activeLeaderTerm field operations + independence from highestLeadershipTerm |
| `StoreIngestionTaskTest.java` | 8 cases for `shouldFilterStaleLeaderRecord` |
| `ActiveActiveStoreIngestionTaskTest.java` | Verify `activeLeaderTerm` is stamped into the chunked-value manifest's `leaderMetadataFooter.termId` |

### Design notes

- `activeLeaderTerm` is intentionally kept separate from `highestLeadershipTerm` so a follower observing a higher term via a DoL stamp from another leader cannot contaminate its own stamping term if it later becomes leader.
- L→S transition: `activeLeaderTerm` is cleared **after** `waitForAllMessageToBeProcessedFromTopicPartition` in the remote-leader-topic branch so any in-flight produces during demotion still carry the demoting leader's real term. The local-VT-only-leader branch never produced from a leader topic, so the clear is done immediately there.
- Thread safety: `shouldFilterStaleLeaderRecord` and `checkAndHandleDoLMessage` both run on the same drainer thread per partition (`StoreBufferService` guarantee), so the read-then-write on `highestLeadershipTerm` is safe without synchronization. Documented in the method javadoc.
- Only data messages are filtered. Control messages (DoL stamps, SOS, EOS, heartbeats) must always be processed: DoL stamps update `highestLeadershipTerm`, SOS/EOS are required for DIV segment tracking, and stale heartbeats are benign.
- `DEFAULT_TERM_ID` (-1) bypass: once all servers are upgraded, this becomes a no-op for new standalone records and manifests but still protects against (1) historical records with -1 that might be replayed, and (2) chunk messages which by design carry `DEFAULT_TERM_ID`.
- Offset/identity advancement on filter: for filtered records we call `pcs.setLatestProcessedVtPosition(...)` directly and return. We deliberately skip the post-process `updateLatestInMemoryProcessedOffset(..., false)` so upstream RT / remote-VT positions, leader GUID, and leader host are not mutated from a stale leader's record. The next non-filtered record will re-run the rewind dry-run check.

### Rollout strategy

1. **Deploy** → All servers start propagating real termIds (Part 1). Filtering remains disabled.
2. **Validate** → Monitor `ingestion.stale_leader_records_filtered.count` (OTel), verify termIds flow correctly.
3. **Enable** → Set `server.leader.term.filtering.enabled=true` per cluster. This is a static config; requires a server restart.

## Test plan

- [x] `PartitionConsumptionStateTest`: `activeLeaderTerm` field operations + independence from `highestLeadershipTerm` (2 tests)
- [x] `StoreIngestionTaskTest`: Full coverage of `shouldFilterStaleLeaderRecord` (8 tests):
  - Filtering disabled → never filters
  - Null `LeaderMetadata` → passes through
  - `DEFAULT_TERM_ID` (-1) → backward-compat pass-through
  - Higher term → updates `highestLeadershipTerm`, passes
  - Equal term → passes
  - Stale term → filtered
  - Stale term → OTel versioned counter emitted via `versionedIngestionStats`
  - End-to-end sequence with real PCS (higher / equal / stale / DEFAULT transitions)
- [x] `ActiveActiveStoreIngestionTaskTest.testLeaderCanSendValueChunksIntoDrainer`: verifies `activeLeaderTerm` from PCS is stamped into the chunked-value manifest's `leaderMetadataFooter.termId`
- [x] `spotlessApply` passes
- [x] `spotbugsMain` passes
- [x] `compileJava` + `compileTestJava` pass

## Follow-up

- Plumb termId through `VeniceWriter.putLargeValue` so chunk messages also carry the producing leader's term. This would let followers filter chunks alongside their manifest rather than leaving orphan chunk bytes in storage when a manifest is filtered.
- Migration-paused S→L path (`isMigrationDuplicateStore`): currently skips both `setActiveLeaderTerm` and `initializeAndSendDoLStamp`. When the transition resumes in `checkLongRunningTaskState`, the `checker` object is no longer available. This is a rare edge case (only during active store migration) and will be addressed in a follow-up PR by persisting the leadership term in PCS during the pause.
- Reduce test reflection fragility in `StoreIngestionTaskTest` via a `@VisibleForTesting` package-private constructor or factory; follows existing pattern in the test class.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java [hunk 1]
```diff
@@ -144,11 +144,24 @@ enum LatchStatus {
...
+  /**
+   * This replica's own leadership term when it is the active leader, used for stamping
+   * outgoing messages. Set during STANDBY_TO_LEADER transition and cleared back to
+   * DEFAULT_TERM_ID during LEADER_TO_STANDBY transition.
+   *
+   * <p>This is intentionally separate from {@link #highestLeadershipTerm}, which tracks the
+   * max term observed from ANY leader (for follower-side filtering). A replica must not use
+   * highestLeadershipTerm for stamping because observing another leader's higher term via a
+   * DoLStamp would incorrectly tag this replica's data with that term.
+   */
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java:164]: Nice design decision keeping `activeLeaderTerm` separate from `highestLeadershipTerm`. The javadoc clearly explains why — a follower observing a higher term via DoLStamp must not contaminate the stamping term. This prevents a subtle bug where a replica that briefly became leader, observed a DoL from a newer leader, then re-became leader would stamp its own data with the newer leader's term.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java [hunk 2]
```diff
@@ -480,6 +480,9 @@ protected void processConsumerAction(ConsumerAction message, Store store) throws
               "State transition from STANDBY to LEADER is paused for replica: {} as this store is undergoing migration",
               partitionConsumptionState.getReplicaId());
         } else {
+          // Set this replica's active leadership term for stamping outgoing messages
+          partitionConsumptionState.setActiveLeaderTerm(checker.getLeadershipTerm());
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java:493]: Minor: `setActiveLeaderTerm` is set before `initializeAndSendDoLStamp` on line 487. If `initializeAndSendDoLStamp` throws, this replica will have a non-default `activeLeaderTerm` set but the transition may not complete. I believe the outer catch would still leave PCS in a bad state regardless, so this is probably fine — but worth a comment noting the ordering is intentional (stamp must be set before the first produce).

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 3]
```diff
@@ -5471,6 +5475,52 @@ void checkAndHandleDoLMessage(PartitionConsumptionState pcs, DefaultPubSubMessag
...
+      return false;
+    }
+
+    long recordTermId = leaderMetadata.getTermId();
+    if (recordTermId == DEFAULT_TERM_ID) {
+      return false;
+    }
+
+    long currentHighest = pcs.getHighestLeadershipTerm();
+    if (recordTermId >= currentHighest) {
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6660]: Thread-safety question: This does a non-atomic read-then-write on `highestLeadershipTerm` (volatile gives visibility but not atomicity). `checkAndHandleDoLMessage` also updates `highestLeadershipTerm` (line ~5389). Are both paths guaranteed to run on the same consumer thread per partition? If yes, a brief comment here noting the single-thread assumption would be helpful. If not, consider `AtomicLong` with `accumulateAndGet(recordTermId, Math::max)` or a CAS loop.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 4]
```diff
@@ -5471,6 +5475,52 @@ void checkAndHandleDoLMessage(PartitionConsumptionState pcs, DefaultPubSubMessag
...
+      if (recordTermId > currentHighest) {
+        pcs.setHighestLeadershipTerm(recordTermId);
+      }
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6668]: Nit: Consider adding a debug log when `highestLeadershipTerm` is updated from a data message, for symmetry with `checkAndHandleDoLMessage` (which logs at INFO). During initial rollout validation, it would help confirm termIds are flowing correctly through the data path:
```java
LOGGER.debug("Replica: {} observed higher termId from data record: {} (previous: {})",
    pcs.getReplicaId(), recordTermId, currentHighest);
```

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 5]
```diff
@@ -3628,6 +3629,9 @@ private int internalProcessConsumerRecord(
           LOGGER.error("Failed to record Record heartbeat with message: ", e);
         }
       } else {
+        if (shouldFilterStaleLeaderRecord(partitionConsumptionState, consumerRecord)) {
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:4370]: This only filters data messages — control messages (heartbeats, SOS, EOS, DoL stamps) take the other branch and are unaffected. Is there a scenario where a stale leader's heartbeat arriving *after* the new leader's data could cause issues? For example, could it confuse heartbeat-delay monitoring or reset consumption state? If intentionally excluded, a brief comment here explaining why would be helpful for future readers.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 6]
```diff
@@ -5471,6 +5475,52 @@ void checkAndHandleDoLMessage(PartitionConsumptionState pcs, DefaultPubSubMessag
...
+    }
+
+    LeaderMetadata leaderMetadata = record.getValue().getLeaderMetadataFooter();
+    if (leaderMetadata == null) {
+      return false;
+    }
+
+    long recordTermId = leaderMetadata.getTermId();
+    if (recordTermId == DEFAULT_TERM_ID) {
+      return false;
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6656]: Good backward-compat safeguard. During mixed rollout, old leaders still produce `DEFAULT_TERM_ID` (-1), and this ensures those records are never filtered. Worth noting that once all servers are upgraded and propagating real termIds (Part 1), this path becomes a no-op for new data — but it still protects against any historical records with -1 that might be replayed.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/HostLevelIngestionStats.java [hunk 7]
```diff
@@ -245,6 +247,12 @@ public HostLevelIngestionStats(
         () -> totalStats.totalOffsetRegressionDCRErrorRate,
         time);
 
+    this.totalStaleLeaderRecordsFilteredRate = registerOnlyTotalRate(
```
Comment (sushantmane) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/HostLevelIngestionStats.java]: Using `registerOnlyTotalRate` means this is a host-level aggregate only (no per-store breakdown at this layer). That makes sense since the per-store granularity comes from `IngestionStats`/`AggVersionedIngestionStats`. Just confirming that's intentional — operators would use the host-level rate for alerting and the versioned stat for drill-down?

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java [hunk 8]
```diff
@@ -7026,4 +7027,199 @@ public void testCheckAndHandleDoLMessageFutureTerm() throws Exception {
...
+   * Helper to set up a mock StoreIngestionTask with the fields needed by shouldFilterStaleLeaderRecord.
+   */
+  private StoreIngestionTask setupMockTaskForFiltering(boolean filteringEnabled) throws Exception {
+    StoreIngestionTask task = mock(StoreIngestionTask.class);
+    doCallRealMethod().when(task).shouldFilterStaleLeaderRecord(any(), any());
+
+    VeniceServerConfig mockServerConfig = mock(VeniceServerConfig.class);
+    doReturn(filteringEnabled).when(mockServerConfig).isLeaderTermFilteringEnabled();
+
+    java.lang.reflect.Field serverConfigField = StoreIngestionTask.class.getDeclaredField("serverConfig");
```
Comment (sushantmane) [inline clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java:7581]: This test helper relies heavily on reflection (`setAccessible(true)`) for `serverConfig`, `versionedIngestionStats`, `storeName`, and `versionNumber`. This is fragile — field renames break silently at runtime, not compile time. Not blocking for this PR since the pattern exists elsewhere in this test class, but worth a follow-up to consider adding a package-private test constructor or `@VisibleForTesting` factory method in `StoreIngestionTask` to inject these dependencies cleanly.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java [hunk 9]
```diff
@@ -7026,4 +7027,199 @@ public void testCheckAndHandleDoLMessageFutureTerm() throws Exception {
...
+    assertTrue(task.shouldFilterStaleLeaderRecord(mockPcs, mockMessage));
+
+    // Verify versioned metric
+    java.lang.reflect.Field versionedStatsField = StoreIngestionTask.class.getDeclaredField("versionedIngestionStats");
+    versionedStatsField.setAccessible(true);
+    AggVersionedIngestionStats versionedStats = (AggVersionedIngestionStats) versionedStatsField.get(task);
+    verify(versionedStats).recordStaleLeaderRecordFiltered("test-store", 1);
+  }
+
+  @Test
```
Comment (sushantmane) [inline clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java:7722]: Excellent end-to-end test using a real PCS instead of mocks. This validates the actual state transitions (higher → equal → stale → DEFAULT → stale) and confirms `highestLeadershipTerm` updates correctly throughout the sequence. This is the most valuable test in the suite.

File: internal/venice-common/src/main/java/com/linkedin/venice/ConfigKeys.java [hunk 10]
```diff
@@ -826,6 +826,18 @@ private ConfigKeys() {
...
 
+  /**
+   * When enabled, followers will filter out stale leader data records whose termId is less than the
+   * highest termId observed so far. This prevents the Delayed Leadership Problem (DLP) where a stale
+   * leader produces to the version topic after a new leader has taken over.
+   *
+   * <p>Records with DEFAULT_TERM_ID (-1) are never filtered, ensuring backward compatibility during
+   * mixed rollout.
+   *
+   * Default: false (filtering disabled; enable after validating termId propagation)
```
Comment (sushantmane) [inline internal/venice-common/src/main/java/com/linkedin/venice/ConfigKeys.java]: Clean javadoc. One thought: would it be useful to document what happens to in-flight records when this flag is toggled from `false` → `true` at runtime (if configs are hot-reloadable)? Or is this strictly a static config that requires a server restart?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java:493]: Good catch. The migration-paused path (`isMigrationDuplicateStore`) indeed skips both `setActiveLeaderTerm` and `initializeAndSendDoLStamp`, and when the transition resumes in `checkLongRunningTaskState`, the `checker` object is no longer available.

Moving `setActiveLeaderTerm` before the if-else alone is not sufficient — `initializeAndSendDoLStamp` also needs to be called at resume time. The proper fix requires storing the leadership term in PCS during the pause so both can be applied when the transition resumes.

This is a rare edge case (only during active store migration), so I will address it in a follow-up PR rather than complicate this one.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java (hunk 1 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/PartitionConsumptionState.java:164]: Thanks\! The separation is critical for correctness — without it, a replica that observes a higher-term DoLStamp from another leader (e.g., during legacy fallback) would incorrectly stamp its own data with that term.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java (hunk 2 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java:493]: Good point. The ordering is intentional — the active leader term must be set before the first produce (DoL stamp) so any messages carry the correct term. Added a comment in 6e4b22b explaining this and noting that the outer exception handler covers the failure case.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 3 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6660]: Both `shouldFilterStaleLeaderRecord` and `checkAndHandleDoLMessage` are called from `internalProcessConsumerRecord`, which runs on the drainer thread. `StoreBufferService` guarantees all records for the same topic+partition go to the same drainer thread, so the read-then-write is safe without additional synchronization.

Added a Javadoc note in 6e4b22b documenting this single-thread-per-partition guarantee.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 4 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6668]: Done in 6e4b22b. Added a `LOGGER.debug` when `highestLeadershipTerm` is updated from a data record. Using debug (not info) since data messages are high volume.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 5 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:4370]: Intentional design. DoL stamps must always be processed — they are the mechanism for updating `highestLeadershipTerm`. SOS/EOS are required for DIV segment tracking. Stale heartbeats are benign: they might briefly make monitoring think things are slightly more recent, but they cannot cause data corruption or false alarms.

Added an inline comment in 6e4b22b explaining the rationale for filtering only data messages.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java (hunk 6 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:6656]: Exactly. Once all servers are upgraded, this path becomes a no-op for new data but still protects against any historical records with -1 that might be replayed from the version topic.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/HostLevelIngestionStats.java (hunk 7 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/HostLevelIngestionStats.java]: Correct — that is the intended split. `registerOnlyTotalRate` at host-level for alerting (total rate across all stores), and the per-store-version counter via `AggVersionedIngestionStats.recordStaleLeaderRecordFiltered()` for drill-down into which store/version is affected.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java (hunk 8 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java:7581]: Agreed — the reflection is fragile. This follows the existing pattern used by other mock-based tests in this class (e.g., `testCheckAndHandleDoLMessage*` tests also use reflection for `localKafkaClusterId`). Will file a follow-up to add a `@VisibleForTesting` package-private constructor or factory method to reduce reflection across all these tests.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java (hunk 9 above)
Comment (sushantmane) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTaskTest.java:7722]: Thanks\! Using a real PCS for the sequence test was intentional — it validates the actual state machine (volatile field updates, cumulative highestLeadershipTerm tracking) rather than just verifying mock interactions.

File: internal/venice-common/src/main/java/com/linkedin/venice/ConfigKeys.java (hunk 10 above)
Comment (sushantmane) [reply internal/venice-common/src/main/java/com/linkedin/venice/ConfigKeys.java]: Good question. This is a static config (`private final boolean`) — it is set once at `VeniceServerConfig` construction time and requires a server restart to change. Added a note to the javadoc in 6e4b22b: "This is a static config; changing it requires a server restart."

**Comment by sushantmane** [comment]
Revived and rebased onto current `main` (was 195 commits behind). Resolved conflicts in `LeaderFollowerStoreIngestionTask` (interaction with upstreamMessageTimestamp + Global RT DIV), `IngestionStats` (uniqueKeyCount), and `StoreIngestionTaskTest` (new SVS / NewStoreVersionState tests). Also updated `sendGlobalRtDivMessage` to stamp `activeLeaderTerm` (new leader-produce path added on main) and adapted PCS test construction to the new 7-arg constructor.

Squashed prior commits (feature + simplification + review-feedback) into a single commit for cleanliness. All earlier review threads addressed and resolved.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java [hunk 11]
```diff
@@ -531,6 +536,9 @@ protected void processConsumerAction(ConsumerAction message, Store store) throws
          *    produce; finally the new follower will switch back to consume from local VT using the latest VT offset
          *    tracked by producer callback.
          */
+        // Clear active leadership term since this replica is no longer the leader
+        partitionConsumptionState.setActiveLeaderTerm(DEFAULT_TERM_ID);
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/LeaderFollowerStoreIngestionTask.java]: Good catch — fixed in 556fe812. Moved the activeLeaderTerm clear in the remote-leader-topic branch to after `waitForAllMessageToBeProcessedFromTopicPartition` so any in-flight produces during demotion still carry the demoting leader's real term. The local-VT-only-leader branch never produced from a leader topic so the clear stays there with a comment noting the rationale.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 12]
```diff
@@ -4230,6 +4231,12 @@ private int internalProcessConsumerRecord(
           LOGGER.error("Failed to record Record heartbeat with message: ", e);
         }
       } else {
+        // Only filter data messages for stale leadership terms. Control messages (heartbeats,
+        // SOS, EOS, DoL stamps) are always processed: DoL stamps are needed to update
+        // highestLeadershipTerm, and SOS/EOS are required for DIV segment tracking.
+        if (shouldFilterStaleLeaderRecord(partitionConsumptionState, consumerRecord)) {
+          return 0;
+        }
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java]: The call sites of `internalProcessConsumerRecord` (via `processConsumerRecord` from `StoreBufferService`) only receive records destined for / read back from the local VT — not RT or remote VT directly. The leader's RT/remote-VT consumer produces transformed records to local VT first; only the local-VT read-back flows through here. DaVinci consumes directly from VT, which is the canonical view, so filtering there is also correct (and it remains opt-in via the disabled-by-default config).

That said, the filter logic does require that any consumed record's `termId` is comparable across the partition's lifetime — i.e., it assumes a single topic per partition feeding the drainer. If a code path ever starts feeding heterogeneous topics through this drainer, the filter would need an explicit local-VT/local-broker gate. I'll document this assumption in the method's javadoc as a follow-up.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 13]
```diff
@@ -4230,6 +4231,14 @@ private int internalProcessConsumerRecord(
...
         }
       } else {
+        /**
+         * Only filter data messages for stale leadership terms. Control messages (heartbeats,
+         * SOS, EOS, DoL stamps) are always processed: DoL stamps are needed to update
+         * highestLeadershipTerm, and SOS/EOS are required for DIV segment tracking.
+         */
+        if (shouldFilterStaleLeaderRecord(partitionConsumptionState, consumerRecord)) {
+          return 0;
+        }
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java:4382]: Good catch — fixed in fb28b4ef. Restructured the data branch to skip `processKafkaDataMessage` for filtered records but fall through to the post-processing block that calls `updateLatestInMemoryProcessedOffset(..., false)` so offsets advance normally. `sizeOfPersistedData` stays 0. This prevents stalling and re-delivery of stale records after restart.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java [hunk 14]
```diff
@@ -38,6 +38,7 @@ public class IngestionStats {
   protected static final String TIMESTAMP_REGRESSION_DCR_ERROR = "timestamp_regression_dcr_error";
   protected static final String OFFSET_REGRESSION_DCR_ERROR = "offset_regression_dcr_error";
   protected static final String TOMBSTONE_CREATION_DCR = "tombstone_creation_dcr";
+  protected static final String STALE_LEADER_RECORDS_FILTERED = "stale_leader_records_filtered";
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/stats/IngestionStats.java]: The implementation is intentionally a rate gauge (`LongAdderRateGauge`), matching the pattern used by every other ingestion event sensor in this class (records consumed, leader records consumed, conflict resolution, batch processing, etc. — all expose events/sec). Operators alert on the rate and drill down with the per-store-version rate when investigating.

Updated the PR description to call it a "rate" rather than "counter". If we want a true monotonically-increasing counter later, that can be added as a separate sensor without changing the existing rate metric.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTaskTest.java [hunk 15]
```diff
@@ -355,6 +361,21 @@ public void testLeaderCanSendValueChunksIntoDrainer() throws InterruptedExceptio
         leaderProducedRecordContextArgumentCaptor.getAllValues().get(3).getKeyBytes(),
         kafkaKeyArgumentCaptor.getAllValues().get(4).getKey());
     verify(mockHostLevelIngestionStats).recordLeaderProduceLatency(anyDouble());
+
+    /**
+     * Verify that the active leader term from PCS is stamped into the leaderMetadataFooter of
+     * the chunked value's manifest message. Chunk messages themselves (indices 1..3) are sent
+     * with DEFAULT_LEADER_METADATA_WRAPPER by VeniceWriter.putLargeValue — only the manifest
+     * (index 4) carries the real termId, because chunks are not standalone records and only
+     * the manifest is what a follower validates against highestLeadershipTerm.
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTaskTest.java:370]: Confirmed and now called out explicitly in the PR description ("Chunked records" section). The PR's protection applies to standalone records and chunked-value manifests; chunk messages keep `DEFAULT_TERM_ID` (preexisting writer behavior). Because a follower can only assemble a chunked value once both chunks and manifest are present, filtering the stale manifest is sufficient to prevent the stale value from being read — the chunks become orphaned bytes in storage but are never assembled. Plumbing `termId` through `VeniceWriter.putLargeValue` so chunks and their manifest are filtered together is listed under "Follow-up" — a clean enhancement but not required for correctness of this PR.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java [hunk 16]
```diff
@@ -4230,17 +4231,29 @@ private int internalProcessConsumerRecord(
...
+         * the drainer from updating latestProcessed* positions, which would re-deliver the same
+         * stale record after restart and stall consumption / RTS progress.
+         */
+        if (!shouldFilterStaleLeaderRecord(partitionConsumptionState, consumerRecord)) {
+          updateLatestInMemoryProcessedOffset(
+              partitionConsumptionState,
+              consumerRecord,
+              leaderProducedRecordContext,
+              kafkaUrl,
+              true);
```
[the remark this answers is not in the document]
Comment (sushantmane) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/StoreIngestionTask.java]: Good catch — fixed in dd15a57e (combined the offset-state isolation with the rate→counter migration the previous review flagged). For filtered records we now call only `pcs.setLatestProcessedVtPosition(consumerRecord.getPosition())` and return, deliberately skipping the post-process `updateLatestInMemoryProcessedOffset(..., false)` so a stale leader's upstream RT / remote-VT positions, leader GUID, and leader host cannot mutate offset tracking. The next non-filtered record will re-run the rewind dry-run check.

I considered also running the dry-run rewind check for filtered records as a safety guard, but skipped it: the rewind check is meaningful when we're about to persist; for a record we're already dropping, it would just be informational noise. Happy to add it back if you'd prefer the extra signal.
